### PR TITLE
Validate fetched layers

### PR DESCRIPTION
**Pull request by hdonnay**
This started out relatively simple, then got away from me.

This changes up the Layer struct, which ends up touching a *lot* of other packages.

This also lays the groundwork for having a unified `Digest` type and adds and refines some testing tools.

### CODE REVIEW

File: layer.go [hunk 1]
```diff
@@ -19,41 +21,144 @@ type RemotePath struct {
 type Layer struct {
 	// content addressable hash unequally identifying this layer. libindex will treat layers with this same
```
Comment (ldelossa) [inline layer.go]: probably my typo but "uniquely"

File: internal/indexer/postgres/manifestscanned_test.go [hunk 2]
```diff
@@ -73,6 +75,8 @@ func Test_ManifestScanned_Success(t *testing.T) {
 	integration.Skip(t)
 	ctx, done := context.WithCancel(context.Background())
 	defer done()
+	const hash = `deafbeef`
```
Comment (ldelossa) [inline internal/indexer/postgres/manifestscanned_test.go]: intended?

Comment (hdonnay) [reply internal/indexer/postgres/manifestscanned_test.go]: yeah, it's just a hex string that's not interpreted by anything in this test.

**Comment by ldelossa** [comment]
@hdonnay 

With this manifest
```
{
    "hash": "0194e59b0b159c8182f2d382953a61d169aa285fe847121f8fc59e7c5a47791a", 
    "layers": [
        {
            "hash": "4ae16bd4778367b46064f39554128dd2fda2803a5747fddeff74059f353391c9", 
            "uri": "https://storage.googleapis.com/quay-sandbox-01/datastorage/registry/sha256/4a/4ae16bd4778367b46064f39554128dd2fda2803a5747fddeff74059f353391c9"
        }, 
        {
            "hash": "a8911d5a30ec53a2c97625050af078e3d57088032981d2eb57b5be0b68f6b5eb", 
            "uri": "https://storage.googleapis.com/quay-sandbox-01/datastorage/registry/sha256/a8/a8911d5a30ec53a2c97625050af078e3d57088032981d2eb57b5be0b68f6b5eb"
        }
    ]
}```

I receive errors 

```
3:54PM ERR handling scan error component=scan-controller manifest=0194e59b0b159c8182f2d382953a61d169aa285fe847121f8fc59e7c5a47791a state=FetchLayers
3:54PM ERR countered error during scan: failed to fetch layers: fetcher: validation failed: got "660314270d7683a945d0cc83e5561d61f022e7e6731fd2e3fad04a6544bf42d8", expected "4ae16bd4778367b46064f39554128dd2fda2803a5747fddeff74059f353391c9" error="failed to fetch layers: fetcher: validation failed: got \"660314270d7683a945d0cc83e5561d61f022e7e6731fd2e3fad04a6544bf42d8\", expected \"4ae16bd4778367b46064f39554128dd2fda2803a5747fddeff74059f353391c9\"" component=scan-controller manifest=0194e59b0b159c8182f2d382953a61d169aa285fe847121f8fc59e7c5a47791ax
```

**Comment by ldelossa** [comment]
> @hdonnay
> 
> With this manifest
> 
> ```
> {
>     "hash": "0194e59b0b159c8182f2d382953a61d169aa285fe847121f8fc59e7c5a47791a", 
>     "layers": [
>         {
>             "hash": "4ae16bd4778367b46064f39554128dd2fda2803a5747fddeff74059f353391c9", 
>             "uri": "https://storage.googleapis.com/quay-sandbox-01/datastorage/registry/sha256/4a/4ae16bd4778367b46064f39554128dd2fda2803a5747fddeff74059f353391c9"
>         }, 
>         {
>             "hash": "a8911d5a30ec53a2c97625050af078e3d57088032981d2eb57b5be0b68f6b5eb", 
>             "uri": "https://storage.googleapis.com/quay-sandbox-01/datastorage/registry/sha256/a8/a8911d5a30ec53a2c97625050af078e3d57088032981d2eb57b5be0b68f6b5eb"
>         }
>     ]
> }
> ```
> 
> I receive errors
> 
> ```
> 3:54PM ERR handling scan error component=scan-controller manifest=0194e59b0b159c8182f2d382953a61d169aa285fe847121f8fc59e7c5a47791a state=FetchLayers
> 3:54PM ERR countered error during scan: failed to fetch layers: fetcher: validation failed: got "660314270d7683a945d0cc83e5561d61f022e7e6731fd2e3fad04a6544bf42d8", expected "4ae16bd4778367b46064f39554128dd2fda2803a5747fddeff74059f353391c9" error="failed to fetch layers: fetcher: validation failed: got \"660314270d7683a945d0cc83e5561d61f022e7e6731fd2e3fad04a6544bf42d8\", expected \"4ae16bd4778367b46064f39554128dd2fda2803a5747fddeff74059f353391c9\"" component=scan-controller manifest=0194e59b0b159c8182f2d382953a61d169aa285fe847121f8fc59e7c5a47791ax
> ```

This was my error, I see that the schema has changed. I attempted manifest scan like this:

```
{
    "hash": "0194e59b0b159c8182f2d382953a61d169aa285fe847121f8fc59e7c5a47791a", 
    "layers": [
        {
            "hash": "sha256:4ae16bd4778367b46064f39554128dd2fda2803a5747fddeff74059f353391c9", 
            "uri": "https://storage.googleapis.com/quay-sandbox-01/datastorage/registry/sha256/4a/4ae16bd4778367b46064f39554128dd2fda2803a5747fddeff74059f353391c9"
        }, 
        {
            "hash": "sha256:a8911d5a30ec53a2c97625050af078e3d57088032981d2eb57b5be0b68f6b5eb", 
            "uri": "https://storage.googleapis.com/quay-sandbox-01/datastorage/registry/sha256/a8/a8911d5a30ec53a2c97625050af078e3d57088032981d2eb57b5be0b68f6b5eb"
        }
    ]
}
```

and received the following error:

```
3:58PM WRN layers validate failure error="encoding/hex: invalid byte: U+0073 's'" manifest=0194e59b0b159c8182f2d382953a61d169aa285fe847121f8fc59e7c5a47791a state=FetchLayers
```
Maye my sha prefix is not correct, or something is off in the hex encoding/decoding

**Pull request merged**